vecs.h

```diff
@@ -0,0 +1,1293 @@
+extern ff_vecs_t skyforce_vecs;
+extern ff_vecs_t pocketfms_bc_vecs;
+extern ff_vecs_t pocketfms_fp_vecs;
+extern ff_vecs_t pocketfms_wp_vecs;
+extern ff_vecs_t v900_vecs;
+extern ff_vecs_t ng_vecs;
+extern ff_vecs_t enigma_vecs;
+extern ff_vecs_t skytraq_vecs;
+extern ff_vecs_t teletype_vecs;
+extern ff_vecs_t skytraq_fvecs;
+extern ff_vecs_t miniHomer_vecs;
+extern ff_vecs_t jogmap_vecs;
+extern ff_vecs_t wintec_tes_vecs;
+extern ff_vecs_t subrip_vecs;
+extern ff_vecs_t format_garmin_xt_vecs;
+extern ff_vecs_t format_fit_vecs;
+extern ff_vecs_t mapbar_track_vecs;
+extern ff_vecs_t f90g_track_vecs;
+extern ff_vecs_t mapfactor_vecs;
+extern ff_vecs_t energympro_vecs;
+extern ff_vecs_t mynav_vecs;
+extern ff_vecs_t geojson_vecs;
+extern ff_vecs_t ggv_bin_vecs;
+extern ff_vecs_t globalsat_sport_vecs;
+#endif // MAXIMAL_ENABLED
+
+class Vecs
+{
+// Meyers Singleton
+public:
+  static Vecs& Instance()
+  {
+    static Vecs instance;
+    return instance;
+  }
+  Vecs(const Vecs&) = delete;
+  Vecs& operator= (const Vecs&) = delete;
+  Vecs(Vecs&&) = delete;
+  Vecs& operator=(Vecs&&) = delete;
+
+private:
+  Vecs() = default;
+  ~Vecs() = default;
+
+private:
+  struct vecs_t {
+    Format* vec;
+    QString name;
+    QString desc;
+    QString extensions; // list of possible extensions separated by '/', first is output default for GUI.
+    QString parent;
+  };
+
+public:
+  void init_vecs();
+  void exit_vecs();
+  static void assign_option(const QString& module, arglist_t* arg, const char* val);
+  static void disp_vec_options(const QString& vecname, const QVector<arglist_t>* args);
+  static void validate_options(const QStringList& options, const QVector<arglist_t>* args, const QString& name);
+  static QString get_option(const QStringList& options, const char* argname);
+  Format* find_vec(const QString& vecname);
+  void disp_vecs() const;
+  void disp_vec(const QString& vecname) const;
+  static const char* name_option(uint32_t type);
+  void disp_formats(int version) const;
+  static bool validate_args(const QString& name, const QVector<arglist_t>* args);
+  bool validate_formats() const;
+
+private:
+  static int is_integer(const char* c);
+  QVector<vecs_t> sort_and_unify_vecs() const;
+  static void disp_v1(ff_type t);
+  static void disp_v2(const Format* v);
+  static void disp_help_url(const vecs_t& vec, const arglist_t* arg);
+  static void disp_v3(const vecs_t& vec);
+  static bool validate_vec(const vecs_t& vec);
+
+private:
+  /*
+   * Having these LegacyFormat instances be non-static data members
+   * prevents the static initialization order fiasco because
+   * the static vec that is used to construct a legacy format
+   * instance is guaranteed to have already constructed when an instance
+   * of this class is constructed.
+   */
+#if CSVFMTS_ENABLED
+  LegacyFormat xcsv_fmt {xcsv_vecs};
+#endif // CSVFMTS_ENABLED
+  LegacyFormat geo_fmt {geo_vecs};
+  GpxFormat gpx_fmt;
+  LegacyFormat mag_sfmt {mag_svecs};
```

Comment:
Can the constructor of, say, geo_fmt be in geo.cc instead of here? This looks like a nearly circular include where everything indirectly includes vecs.h and any change to any format results in a recompilation of everything.
Even if each format has a global LegacyFormat geo_fmt{geo_vecs} (globals are bad. i know...) that registered the blahFormat, that would be one less file to modify/recompile when things are changed, right?

Reply:
I think the constructors could be local to each format, but that would require editing every format.  In either case there is a global, e.g. either geo_fmt or geo_vecs.

As it stands a change to a format only requires recompilation of that format.  A change to vecs.h or legacyformat.h requires a few files to be recompiled.  With your suggestion a change to legacyformat.h would require every format to be recompiled.

If we every go to dynamic formats then I don't think the constructor belongs in the format.